### PR TITLE
[Serialization] Pydantic -> `serialization_addons.py` and Ray Client support.

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
This PR bring back the serialization_addon mechanism from https://github.com/ray-project/ray/pull/12478 and use it to register custom serializers for `pydantic` and `starlette`. Additionally, this PR make sure Ray Client will use the same mechanism.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
I assume this closes https://github.com/ray-project/ray/issues/14944?

**Review by edoakes — COMMENTED** [review]
Looks good, please add an end-to-end test Serve test w/ fastapi to the `serve/tests/ray_client_test.py`

**Comment by edoakes** [comment]
cc @AmeerHajAli

File: python/ray/util/client/ray_client_helpers.py [hunk 1]
```diff
@@ -41,3 +42,20 @@ def ray_connect_handler(job_config=None):
...
         ray.disconnect()
         server.stop(0)
+
+
+class RayClientSerializationContext:
+    # NOTE(simon): Used for registering custom serializers. We cannot directly
+    # use the SerializationContext because it requires Ray workers. Please
+    # make sure to keep the API consistent.
+
+    def _unregister_cloudpickle_reducer(self, cls):
```
Comment (ijrsvt) [inline python/ray/util/client/ray_client_helpers.py:52]: Hmm, do we need any of the other methods for this? 

I think this approach is fine for now, but I wonder if a better long term goal to have a common base SerializationContext class and then have a more specialized one for workers.

File: python/ray/serialization_addons.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+def register_pydantic_serializer(serialization_context):
+    try:
+        import pydantic.fields
+    except ImportError:
+        return
+
+    # Pydantic's Cython validators are not serializable.
+    # https://github.com/cloudpipe/cloudpickle/issues/408
+    serialization_context._register_cloudpickle_serializer(
```
Comment (robertnishihara) [inline python/ray/serialization_addons.py:15]: What's the reason for using an internal API instead of `ray.util.register_serializer`?

**Comment by AmeerHajAli** [comment]
Ping @simon-mo

File: python/ray/serialization_addons.py (hunk 2 above)
Comment (simon-mo) [reply python/ray/serialization_addons.py:15]: The ray.util method doesn't work for Ray client's worker.

Comment (edoakes) [reply python/ray/serialization_addons.py:15]: We should make a unified public API! @AmeerHajAli can you make sure to follow up on this?

**Review by edoakes — APPROVED** [review]

File: python/ray/serialization_addons.py (hunk 2 above)
Comment (AmeerHajAli) [reply python/ray/serialization_addons.py:15]: CC @ijrsvt who is leading these efforts.

**Comment by simon-mo** [comment]
windows:test_runtime_env fails in master
buildkite: tests:test_placement_group & tests:custom_env_torch are known flaky/failing as well

**Review by ijrsvt — COMMENTED** [review]
Looks good! thanks for the test!!

**Pull request merged**